### PR TITLE
Add Xbox Support for Java

### PR DESCRIPTION
**Pull request by cpapplefamily**
This has all the edits required to add Xbox Controller and Xbox buttons to the menus and output Java code. Also Allows for future "Joystick" type Extensions to be created.

### CODE REVIEW

**Comment by sciencewhiz** [comment]
Can you rebase this on master? It doesn't change cleanly if I change the PR to master.

**Comment by cpapplefamily** [comment]
> Can you rebase this on master? It doesn't change cleanly if I change the PR to master.

Is that simply using the following while on my branch?
`git pull --rebase upstream master`

Then push to my online repo

**Comment by sciencewhiz** [comment]
Yes. You will get merge conflicts with some of the duplicate commits that you will need to skip. You will also need to force push to your branch.

File: src/main/resources/help/Xbox Controller.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        href="styles.css"
+        type="text/css"
+        media="screen" />
+  <title>Xbox Controller</title>
+</head>
+<body>
+  <h1>Xbox Controller</h1>
+  <center>
+    <img src="../icons/Xbox Controller.png"
+            alt="Joystick" />
```
Comment (sciencewhiz) [inline src/main/resources/help/Xbox Controller.html:17]:
```suggestion
            alt="Xbox Controller" />
```

**Review by sciencewhiz — CHANGES_REQUESTED** [review]

**Comment by sciencewhiz** [comment]
Use the Joystick Button html from #234 as a basis

**Comment by cpapplefamily** [comment]
> Yes. You will get merge conflicts with some of the duplicate commits that you will need to skip. You will also need to force push to your branch.

I only see the one conflict Should I select Accept Incoming? 
Capture

**Comment by sciencewhiz** [comment]
You want to accept current change for that one. I suspect once you clear that one, there will be more.

**Comment by cpapplefamily** [comment]
> Yes. You will get merge conflicts with some of the duplicate commits that you will need to skip. You will also need to force push to your branch.

@sciencewhiz 
Is this supposed to be a pain in the backside? I have tried to suffer through and honestly would have been easier to create a new branch and re-apply my edits. I would love to watch the process if someone wanted to screen share.

**Comment by cpapplefamily** [comment]
Commit two after the first time I `git rebase --continue`
My steps

1. Check out Xbox-Controller-Java-Support
2. `git pull --rebase origin master`  (*origin is equal to upstream now)
3. Fix conflict or `accept current change`
4.  `git add src/main/java/robotbuilder/exporters/GenericExporter.java`
5. `git rebase --continue` This is where it goes south and all my attempts usually end here:

```
E325: ATTENTION
Found a swap file by the name "C:/Users/Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/.COMMIT_EDITMSG.swp"
          owned by: Robotics   dated: Wed Oct 28 08:23:57 2020
         file name: ~Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/COMMIT_EDITMSG
          modified: no
         user name: Robotics   host name: Robotics-PC
        process ID: 184 (STILL RUNNING)
While opening file "C:/Users/Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/COMMIT_EDITMSG"
             dated: Wed Oct 28 18:50:07 2020
      NEWER than swap file!

(1) Another program may be editing the same file.  If this is the case,
    be careful not to end up with two different instances of the same
    file when making changes.  Quit, or continue with caution.
(2) An edit session for this file crashed.
    If this is the case, use ":recover" or "vim -r C:/Users/Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/COMMIT_EDITMSG"
    to recover the changes (see ":help recovery").
    If you did this already, delete the swap file "C:/Users/Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/.COMMIT_EDITMSG.swp"
    to avoid this message.

Swap file "C:/Users/Robotics/FRCProjects 2021/WPILIB_RobotBuilder/RobotBuilder/.git/.COMMIT_EDITMSG.swp" already exists!
[O]pen Read-Only, (E)dit anyway, (R)ecover, (Q)uit, (A)bort: 
```

**Comment by sciencewhiz** [comment]
I'm not sure why you're getting vim swap file errors, unless its the combination of vs code and git commandline.

cherry-picking on a clean branch is also an option.

**Comment by cpapplefamily** [comment]
Maybe Ill try to use the Git-SVN outside VSCode

**Comment by cpapplefamily** [comment]
@sciencewhiz I either 100% was successful or not. It the rebase is not correct I will delete my branch and start a new one off the current Master. Do I need to change the into branch? because you had me rebase to the master.

**Comment by sciencewhiz** [comment]
After changing the PR base branch to master, it looks much better. However, it looks like there's unitentional changes to command.java.

**Comment by cpapplefamily** [comment]
I recall that one giving me some trouble. If that's all then great and I'll be a bit more careful on the next one.

File: src/main/resources/export/java/macros.vm [hunk 2]
```diff
@@ -7,6 +7,9 @@
 ## Generate Java style consant names from a component name
 #macro( constant $name )${name.replace(' ', '_').toUpperCase()}#end
 
+## Get the type Not ClassName of a robot component
+#macro( type2 $component )$component.getBase().getType()#end
```
Comment (sciencewhiz) [inline src/main/resources/export/java/macros.vm:11]: I would prefer a more descriptive name, like base_type

**Comment by cpapplefamily** [comment]
Sure I get it. That was obviously a test macro to attempt to get the correct data and not break other calls to it. As a follow-up should 
`#macro( type $component )${helper.getInstruction($component, "ClassName")}#end`
Get refactored to something that is more representative to get_class? Probably more involved and worthy of a new issue tracker.

**Comment by sciencewhiz** [comment]
I think type is fine as the class is the type

**Comment by cpapplefamily** [comment]
Suggested change complete

**Comment by sciencewhiz** [comment]
superseded by #335

**Pull request closed without merging**